### PR TITLE
fix(rds): instance identifiers and endpoints of a Cluster are blank

### LINKED ISSUE #14377: aws-rds: instance identifiers and endpoints not available on DatabaseCluster
DatabaseCluster should have instanceIdentifiers and instanceEndpoints props available. However, it looks like these are always empty.

From the code we can see that those are not actually set to the right values returned by the `createInstances` method:
https://github.com/aws/aws-cdk/blob/55e737ffb378cc8690b201a6c52424102dee9c3b/packages/%40aws-cdk/aws-rds/lib/cluster.ts#L527

### Reproduction Steps

Create a DatabaseCluster with at least 1 instance and try to access instanceIdentifiers or instanceEndpoints prop. Those are always empty arrays.

This is :bug: Bug Report

### PR DESCRIPTION
**Pull request by vilikin**
Previously instanceIdentifiers and instanceEndpoints were readonly properties of DatabaseClusterNew, defaulting to empty arrays. However, they they were never set to the correct values after instances were added to the cluster. Those properties are now moved on to the DatabaseCluster and DatabaseClusterFromSnapshot classes where they can now actually be set to correct values.

fixes #14377

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-rds/lib/cluster.ts [hunk 1]
```diff
@@ -290,8 +290,6 @@ abstract class DatabaseClusterNew extends DatabaseClusterBase {
    * Never undefined.
    */
   public readonly engine?: IClusterEngine;
-  public readonly instanceIdentifiers: string[] = [];
-  public readonly instanceEndpoints: Endpoint[] = [];
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:294]: Why move these? We keep members in a specific order (public, then protected, then private). Let's keep it this way here.

File: packages/@aws-cdk/aws-rds/lib/cluster.ts [hunk 2]
```diff
@@ -484,6 +482,9 @@ export class DatabaseCluster extends DatabaseClusterNew {
   private readonly singleUserRotationApplication: secretsmanager.SecretRotationApplication;
   private readonly multiUserRotationApplication: secretsmanager.SecretRotationApplication;
 
+  public readonly instanceIdentifiers: string[] = [];
+  public readonly instanceEndpoints: Endpoint[] = [];
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:486]: Let's not have these have default values. They're not used anyway, and without them the compiler will make sure these get initialized. So:

```suggestion
  public readonly instanceIdentifiers: string[];
  public readonly instanceEndpoints: Endpoint[];
```

File: packages/@aws-cdk/aws-rds/lib/cluster.ts [hunk 3]
```diff
@@ -524,7 +525,9 @@ export class DatabaseCluster extends DatabaseClusterNew {
     }
 
     setLogRetention(this, props);
-    createInstances(this, props, this.subnetGroup);
+    const { instanceIdentifiers, instanceEndpoints } = createInstances(this, props, this.subnetGroup);
+    this.instanceIdentifiers = instanceIdentifiers;
+    this.instanceEndpoints = instanceEndpoints;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:530]: Super minor, but I feel like this is a little bit repetitive. How about

```suggestion
    const createdInstances = createInstances(this, props, this.subnetGroup);
    this.instanceIdentifiers = createdInstances.instanceIdentifiers;
    this.instanceEndpoints = createdInstances.instanceEndpoints;
```

?

File: packages/@aws-cdk/aws-rds/lib/cluster.ts [hunk 4]
```diff
@@ -594,6 +597,9 @@ export class DatabaseClusterFromSnapshot extends DatabaseClusterNew {
   public readonly clusterReadEndpoint: Endpoint;
   public readonly connections: ec2.Connections;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:599]: I don't think there's a need for this empty line here.

File: packages/@aws-cdk/aws-rds/lib/cluster.ts [hunk 5]
```diff
@@ -594,6 +597,9 @@ export class DatabaseClusterFromSnapshot extends DatabaseClusterNew {
...
 
+  public readonly instanceIdentifiers: string[] = [];
+  public readonly instanceEndpoints: Endpoint[] = [];
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:601]: Same comment here as above (get rid of the initializers).

File: packages/@aws-cdk/aws-rds/lib/cluster.ts [hunk 6]
```diff
@@ -616,7 +622,9 @@ export class DatabaseClusterFromSnapshot extends DatabaseClusterNew {
     cluster.applyRemovalPolicy(props.removalPolicy ?? RemovalPolicy.SNAPSHOT);
 
     setLogRetention(this, props);
-    createInstances(this, props, this.subnetGroup);
+    const { instanceIdentifiers, instanceEndpoints } = createInstances(this, props, this.subnetGroup);
+    this.instanceIdentifiers = instanceIdentifiers;
+    this.instanceEndpoints = instanceEndpoints;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:627]: Same comment here as above (no need for the destructuring form).

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts [hunk 7]
```diff
@@ -1902,6 +1902,77 @@ describe('cluster', () => {
...
+            },
+            ':',
+            {
+              'Fn::GetAtt': ['DatabaseB269D8BB', 'Endpoint.Port'],
+            },
+          ],
+        ],
+      },
+    });
+  });
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/cluster.test.ts:1975]: This is such a small change, I'm not sure it warrants an entirely separate unit test.

How about adding a new assertion to existing tests?

**Review by skinny85 — CHANGES_REQUESTED** [review]
Thanks for the contribution @vilikin! It looks great. However, since I'm a picky asshole, I've managed to find a few places I would do things differently. I hope you agree with my suggestions!

File: packages/@aws-cdk/aws-rds/lib/cluster.ts (hunk 1 above)
Comment (vilikin) [reply packages/@aws-cdk/aws-rds/lib/cluster.ts:294]: They are actually moved to the implementing classes, so they no longer exist on the `DatabaseClusterNew` class. (diff looks almost as if they were just moved a bit down inside same class, but there's actually 200 rows in between)

The reason for moving them was that I could not figure out a way to keep them readonly on the base class if we needed to set them from the implementing classes.

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts (hunk 7 above)
Comment (vilikin) [reply packages/@aws-cdk/aws-rds/test/cluster.test.ts]: I now included the expectations to an existing unit test that was creating a cluster with 1 db instance.

But I do feel like the existing unit tests have been pretty specific on testing just a certain functionality, so it didn't fit too naturally into any existing test in my opinion.

**Comment by vilikin** [comment]
> Thanks for the contribution @vilikin! It looks great. However, since I'm a picky asshole, I've managed to find a few places I would do things differently. I hope you agree with my suggestions!

Thanks for the thorough review, I do appreciate even the nitpicks :)

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts [hunk 8]
```diff
@@ -95,6 +95,35 @@ describe('cluster', () => {
...
+          [
+            {
+              'Fn::GetAtt': ['DatabaseInstance1844F58FD', 'Endpoint.Address'],
+            },
+            ':',
+            {
+              'Fn::GetAtt': ['DatabaseB269D8BB', 'Endpoint.Port'],
+            },
+          ],
+        ],
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/cluster.test.ts:124]: Let's maybe save some lines here:

```suggestion
        'Fn::Join': ['', [
          { 'Fn::GetAtt': ['DatabaseInstance1844F58FD', 'Endpoint.Address'] },
          ':',
          { 'Fn::GetAtt': ['DatabaseB269D8BB', 'Endpoint.Port'] },
        ]],
```

File: packages/@aws-cdk/aws-rds/lib/cluster.ts [hunk 9]
```diff
@@ -484,6 +482,9 @@ export class DatabaseCluster extends DatabaseClusterNew {
   private readonly singleUserRotationApplication: secretsmanager.SecretRotationApplication;
   private readonly multiUserRotationApplication: secretsmanager.SecretRotationApplication;
 
+  public readonly instanceIdentifiers: string[];
+  public readonly instanceEndpoints: Endpoint[];
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:486]: > They are actually moved to the implementing classes, so they no longer exist on the `DatabaseClusterNew` class. (diff looks almost as if they were just moved a bit down inside same class, but there's actually 200 rows in between)
> 
> The reason for moving them was that I could not figure out a way to keep them readonly on the base class if we needed to set them from the implementing classes.

Ah, right. I missed that. Sounds good!

Can you please move these to the block of the `public` fields above? (The one that ends with `  public readonly connections: ec2.Connections;`).

**Review by skinny85 — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-rds/lib/cluster.ts (hunk 9 above)
Comment (vilikin) [reply packages/@aws-cdk/aws-rds/lib/cluster.ts]: Oops, forgot to address that part of the comment. Now it should be ok :)

File: packages/@aws-cdk/aws-rds/lib/cluster.ts [hunk 10]
```diff
@@ -484,6 +484,7 @@ export class DatabaseCluster extends DatabaseClusterNew {
   private readonly singleUserRotationApplication: secretsmanager.SecretRotationApplication;
   private readonly multiUserRotationApplication: secretsmanager.SecretRotationApplication;
 
+
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/lib/cluster.ts:487]: No need for this empty line.

File: packages/@aws-cdk/aws-rds/test/cluster.test.ts [hunk 11]
```diff
@@ -95,6 +95,28 @@ describe('cluster', () => {
...
+        'Fn::Join': ['', [
+          { 'Fn::GetAtt': ['DatabaseInstance1844F58FD', 'Endpoint.Address'] },
+          ':',
+          { 'Fn::GetAtt': ['DatabaseB269D8BB', 'Endpoint.Port'] },
+        ]],
+      },
+    });
   });
 
   test('can create a cluster with imported vpc and security group', () => {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-rds/test/cluster.test.ts:122]: We're missing a test for the `DatabaseClusterFromSnapshot`.

**Review by skinny85 — COMMENTED** [review]
2 small things are still not perfect - any chance of getting these in @vilikin?

**Review by skinny85 — APPROVED** [review]
Awesome contribution. Thanks for enduring my pickiness @vilikin 🙂.

**Comment by skinny85** [comment]
The build fails with

```
@aws-cdk/aws-rds: FAIL test/cluster.test.js (14.981 s)
@aws-cdk/aws-rds:   â— cluster â€º create a cluster from a snapshot
@aws-cdk/aws-rds:     expect(received).toContain(expected) // indexOf
@aws-cdk/aws-rds:     Expected value:  {"Ref": "DatabaseInstance1844F58FD"}
@aws-cdk/aws-rds:     Received object: {"Ref": "DatabaseInstance1844F58FD"}
```

I think this

```ts
    expect(stack.resolve(cluster.instanceIdentifiers[0])).toContain({
      Ref: 'DatabaseInstance1844F58FD',
    });
```

should be

```ts
    expect(stack.resolve(cluster.instanceIdentifiers[0])).toEqual({
      Ref: 'DatabaseInstance1844F58FD',
    });
```

As in the test for `DatabaseCluster`.

**Comment by vilikin** [comment]
> Awesome contribution. Thanks for enduring my pickiness @vilikin 🙂.

Thanks for all the comments, they were all very much valid. I was just a bit careless due to lack of time. (which I obviously then spent more in the end... :D)

Is this now just going to be auto-merged at some point or is there still something to do?

**Comment by vilikin** [comment]
> The build fails with
> 
> ```
> @aws-cdk/aws-rds: FAIL test/cluster.test.js (14.981 s)
> @aws-cdk/aws-rds:   â—� cluster â€º create a cluster from a snapshot
> @aws-cdk/aws-rds:     expect(received).toContain(expected) // indexOf
> @aws-cdk/aws-rds:     Expected value:  {"Ref": "DatabaseInstance1844F58FD"}
> @aws-cdk/aws-rds:     Received object: {"Ref": "DatabaseInstance1844F58FD"}
> ```
> 
> I think this
> 
> ```ts
>     expect(stack.resolve(cluster.instanceIdentifiers[0])).toContain({
>       Ref: 'DatabaseInstance1844F58FD',
>     });
> ```
> 
> should be
> 
> ```ts
>     expect(stack.resolve(cluster.instanceIdentifiers[0])).toEqual({
>       Ref: 'DatabaseInstance1844F58FD',
>     });
> ```
> 
> As in the test for `DatabaseCluster`.

Ooops, yep. That's now fixed.

**Review by skinny85 — APPROVED** [review]
> > The build fails with
> > ```
> > @aws-cdk/aws-rds: FAIL test/cluster.test.js (14.981 s)
> > @aws-cdk/aws-rds:   â—� cluster â€º create a cluster from a snapshot
> > @aws-cdk/aws-rds:     expect(received).toContain(expected) // indexOf
> > @aws-cdk/aws-rds:     Expected value:  {"Ref": "DatabaseInstance1844F58FD"}
> > @aws-cdk/aws-rds:     Received object: {"Ref": "DatabaseInstance1844F58FD"}
> > ```
> > 
> > 
> > I think this
> > ```ts
> >     expect(stack.resolve(cluster.instanceIdentifiers[0])).toContain({
> >       Ref: 'DatabaseInstance1844F58FD',
> >     });
> > ```
> > 
> > 
> > should be
> > ```ts
> >     expect(stack.resolve(cluster.instanceIdentifiers[0])).toEqual({
> >       Ref: 'DatabaseInstance1844F58FD',
> >     });
> > ```
> > 
> > 
> > As in the test for `DatabaseCluster`.
> 
> Ooops, yep. That's now fixed.

Awesome, thanks so much!

**Comment by skinny85** [comment]
@vilikin looks like there's a merge conflict with the current master. Do you have time to resolve it?

**Comment by vilikin** [comment]
> @vilikin looks like there's a merge conflict with the current master. Do you have time to resolve it?

Actually, this time I really don't have time anymore. We are expecting a baby and the water just broke, so we are off to the hospital soon - so if you'd be able to resolve it, I would highly appreciate it. :D

**Comment by skinny85** [comment]
> > @vilikin looks like there's a merge conflict with the current master. Do you have time to resolve it?
> 
> Actually, this time I really don't have time anymore. We are expecting a baby and the water just broke, so we are off to the hospital soon - so if you'd be able to resolve it, I would highly appreciate it. :D

That is a _great_ excuse for not resolving merge conflicts. I'll handle this one. Best of luck with the baby!

**Review by skinny85 — APPROVED** [review]

**Pull request merged**